### PR TITLE
Js complete goals

### PR DESCRIPTION
**Pull request by johnschoeman**
This allows users to toggle goals to indicate if they have completed them or not.

---

### Before: 

togglegoals - before

### After: 

togglegoals - after

---

Had to introduce babel and polyfills for fetch and promises to be able to write tests using capybara and webkit.  the commits are not squished down because I'd like to spend some time making notes on the journey I took to get capybara to play nice with es6 and the fetch api.  I think that this has potential to make a good blog post/tutorial.

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -289,13 +300,15 @@ DEPENDENCIES
   clearance
   delayed_job_active_record
   dotenv-rails
+  es6-promise-rails
```
Comment (johnschoeman) [inline Gemfile.lock:303]: brought in the es6-promise-rails and fetch-rails gem to get the polyfills. I'd like to try and get the polyfills directly to reduce the number of dependencies, currently a WIP.

File: app/assets/javascripts/goal.es6 [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+      fetch(url, {
+        method: 'PUT',
+        headers: new Headers({
+          'X-Requested-With': 'XMLHttpRequest',
+          'X-CSRF-Token': csrfToken
+        }),
+        credentials: 'same-origin'
+      }).then(res => handleErrors(res))
+        .catch(err => console.log(err));
```
Comment (stevehanson) [inline app/assets/javascripts/goal.es6]: It might be a good idea at some point to update the UI when the request fails to make sure the checkbox state is consistent with the server state. As-is, the checkbox would remain checked/unchecked, even if the operation failed on the server.

File: app/assets/javascripts/goal.es6 [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+    }
+  }, false);
+})
+
+const handleErrors = (res) => {
+  if (!res.ok) {
+    const alertMessage = `Error saveing goal: status: ${res.status} ${res.statusText}`;
```
Comment (stevehanson) [inline app/assets/javascripts/goal.es6]: When there is a network issue (as opposed to a failed 4xx/5xx request), `fetch` fails the promise, so this piece of code won't be hit. It seems like it might make sense to just throw an error here:

```es6
if (!res.ok) {
  throw Error(res.statusText);
}
```

and then do the `alert` in the promise `catch`. That way, the alert will be shown even when network errors occur.

File: app/views/entries/show.html.erb [hunk 4]
```diff
@@ -9,7 +9,14 @@
   <h2>Goals:</h2>
   <ul class="goal-list">
   <% @entry.goals.each do |goal| %>
-    <li class="goal-summary"><%= goal.description %></li>
+    <% goal_id = "goal-#{goal.id}" %>
+    <li class="goal-summary" >
+      <label class="goal-checkbox-label">
+        <%= check_box_tag goal_id, true, checked = goal.completed,
```
Comment (stevehanson) [inline app/views/entries/show.html.erb]: You could change `checked = goal.completed` to just `goal.completed`.

File: spec/features/user_completes_a_goal_spec.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
+require "rails_helper"
+require "support/features/clearance_helpers"
+
+RSpec.feature "User completes a goal" do
+  scenario "User clicks an uncomplete goal", js: true do
+    user = create(:user)
+    entry = create(:entry, user: user)
+    goal1_description = Faker::Lorem.sentence
+    goal1 = Goal.create(entry: entry,
```
Comment (stevehanson) [inline spec/features/user_completes_a_goal_spec.rb]: Is there a reason why you opted to not use `factory-bot` here?

File: spec/requests/post_api_goal_request_spec.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
+require "rails_helper"
+require "support/features/clearance_helpers"
+
+RSpec.describe "PATCH /api/goals/:id", type: :request do
```
Comment (stevehanson) [inline spec/requests/post_api_goal_request_spec.rb:4]: Nice specs!

File: .bablerc [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (stevehanson) [inline .bablerc]: Should this file be called `.babelrc`?

**Review by stevehanson — APPROVED** [review]
Great PR! 💯 I just left a few small thoughts.

File: .bablerc (hunk 7 above)
Comment (johnschoeman) [reply .bablerc]: it should be, good catch!

File: spec/requests/post_api_goal_request_spec.rb (hunk 6 above)
Comment (johnschoeman) [reply spec/requests/post_api_goal_request_spec.rb:4]: TY!

File: spec/features/user_completes_a_goal_spec.rb (hunk 5 above)
Comment (johnschoeman) [reply spec/features/user_completes_a_goal_spec.rb]: no, I will make the revision.

File: app/assets/javascripts/goal.es6 (hunk 3 above)
Comment (johnschoeman) [reply app/assets/javascripts/goal.es6]: 💯

File: app/assets/javascripts/goal.es6 [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        headers: new Headers({
+          'X-Requested-With': 'XMLHttpRequest',
+          'X-CSRF-Token': csrfToken
+        }),
+        credentials: 'same-origin'
+      }).then(res => handleErrors(res))
+        .catch(err => console.log(err));
+
+    }
+  }, false);
```
Comment (christoomey) [inline app/assets/javascripts/goal.es6]: Can this `fetch` call be extracted into another function?

File: app/assets/javascripts/goal.es6 [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+      let csrfToken = document.querySelector('meta[name=csrf-token]').content;
+      fetch(url, {
+        method: 'PUT',
+        headers: new Headers({
+          'X-Requested-With': 'XMLHttpRequest',
+          'X-CSRF-Token': csrfToken
+        }),
+        credentials: 'same-origin'
+      }).then(res => handleResult(res))
+        .catch(err => handleErrors(err));
```
Comment (christoomey) [inline app/assets/javascripts/goal.es6:16]: These can be shortened to `.then(handleResult)` and `.catch(handleErrors);`. This is referred to as [point-free style](https://stackoverflow.com/questions/944446/what-is-point-free-style-in-functional-programming?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa). 

This also works with things like `console.log` and `console.error`, which is nifty.

File: app/controllers/api/goals_controller.rb [hunk 10]
```diff
@@ -0,0 +1,26 @@
+module Api
+  class GoalsController < ApplicationController
+    before_action :require_entry_to_belong_to_user
+
+    def update
+      goal.toggle(:completed)
+      goal.save
```
Comment (christoomey) [inline app/controllers/api/goals_controller.rb:7]: I'd shorten this to `goal.update!(completed: true)` or perhaps add a method like `goal.toggleComplete!`. The `!` isn't required if you're handling the `false` case of `save` or `update`.

Also, I'd recommend returning something here, although it can be as simple as `head :ok` which is just a status 200 response with no content.

File: app/controllers/api/goals_controller.rb [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+      @_goal ||= Goal.find(params[:id])
+    end
+
+    def entry
+      @_entry_for_goal ||= goal.entry
+    end
+
+    def require_entry_to_belong_to_user
+      unless entry.user == current_user
+        head(403)
```
Comment (christoomey) [inline app/controllers/api/goals_controller.rb:22]: What is this magical sorcery number 403?!?! I think you can use `head :forbidden` or similar here (I forget the exact symbol, but I'm pretty sure there is one). :)

**Review by christoomey — COMMENTED** [review]
A couple of comments that are by no means critical, but worth highlighting / chatting about.
